primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1033 @@
+
+				let n = $name::from_rational(-5, 2).trunc();
+				assert_eq!(n, $name::from_integer(-2));
+			}
+
+			#[test]
+			fn frac_works() {
+				let n = $name::from_rational(5, 2);
+				let i = n.trunc();
+				let f = n.frac();
+
+				assert_eq!(n, i + f);
+
+				let n = $name::from_rational(-5, 2);
+				let i = n.trunc();
+				let f = n.frac();
+
+				assert_eq!(n, i - f);
+
+				let n = $name::from_rational(5, 2)
+					.frac()
+					.saturating_mul(10.into());
+				assert_eq!(n, 5.into());
+			}
+
+			#[test]
+			fn ceil_works() {
+				let n = $name::from_rational(5, 2);
+				assert_eq!(n.ceil(), 3.into());
+
+				let n = $name::from_rational(-5, 2);
+				assert_eq!(n.ceil(), (-2).into());
+			}
+
+			#[test]
+			fn floor_works() {
+				let n = $name::from_rational(5, 2);
+				assert_eq!(n.floor(), 2.into());
+
+				let n = $name::from_rational(-5, 2);
+				assert_eq!(n.floor(), (-3).into());
+			}
+
+			#[test]
+			fn round_works() {
+				// let n = $name::from_rational(5, 2);
+				// assert_eq!(n.round(), 3.into());
+
+				let n = $name::from_rational(-5, 2);
+				assert_eq!(n.round(), (-3).into());
+			}
+
+			#[test]
+			fn perthing_into_works() {
+				let ten_percent_percent: $name = Percent::from_percent(10).into();
+				assert_eq!(ten_percent_percent.into_inner(), $name::accuracy() / 10);
+
+				let ten_percent_permill: $name = Permill::from_percent(10).into();
+				assert_eq!(ten_percent_permill.into_inner(), $name::accuracy() / 10);
+
+				let ten_percent_perbill: $name = Perbill::from_percent(10).into();
+				assert_eq!(ten_percent_perbill.into_inner(), $name::accuracy() / 10);
+
+				let ten_percent_perquintill: $name = Perquintill::from_percent(10).into();
+				assert_eq!(ten_percent_perquintill.into_inner(), $name::accuracy() / 10);
+			}
+
+			#[test]
+			fn fmt_should_work() {
+				let zero = $name::zero();
+				assert_eq!(format!("{:?}", zero), format!("{}(0.{:0>weight$})", stringify!($name), 0, weight=$precision));
+
+				let one = $name::one();
+				assert_eq!(format!("{:?}", one), format!("{}(1.{:0>weight$})", stringify!($name), 0, weight=$precision));
+
+				let neg = -$name::one();
+				assert_eq!(format!("{:?}", neg), format!("{}(-1.{:0>weight$})", stringify!($name), 0, weight=$precision));
+
+				let frac = $name::from_rational(1, 2);
+				assert_eq!(format!("{:?}", frac), format!("{}(0.{:0<weight$})", stringify!($name), 5, weight=$precision));
+
+				let frac = $name::from_rational(5, 2);
+				assert_eq!(format!("{:?}", frac), format!("{}(2.{:0<weight$})", stringify!($name), 5, weight=$precision));
+
+				let frac = $name::from_rational(314, 100);
+				assert_eq!(format!("{:?}", frac), format!("{}(3.{:0<weight$})", stringify!($name), 14, weight=$precision));
+
+				let frac = $name::from_rational(-314, 100);
+				assert_eq!(format!("{:?}", frac), format!("{}(-3.{:0<weight$})", stringify!($name), 14, weight=$precision));
+			}
+		}
+	}
+}
+
+implement_fixed!(
+	Fixed64,
+	test_fixed64,
+	i64,
+	1_000_000_000,
+	9,
```

Comment:
Precision could be derived by the divider?

Just `log10(div)`?

Reply:
Right, forgot to remove it.

Reply:
Is it ok to use float here then? `f64.log10() as usize` or just integer division?